### PR TITLE
Implement enumerable methods

### PR DESCRIPTION
**Pull request by edemagbenyo**
Implement the following methods:
my_each
my_each_with_index
my_all?
my_any?
my_none?
my_count
my_map
my_inject

### CODE REVIEW

File: index.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    arr = to_a
+    arr.to_enum unless block_given?
+    n = 0
```
Comment (samgaco) [inline index.rb:11]: - [ ] my_each returns an Enumerator if no block is given

File: index.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    my_each do |v|
+      res << v if yield(v) == true
+    end
+    res
+  end
+
+  def my_all?
+    check = true
+    if block_given?
+      my_each { |i| check = false unless yield(i) }
```
Comment (samgaco) [inline index.rb]: - [ ] my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: index.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    else
+      my_all? { |obj| obj }
+    end
+    check
+  end
+
+  def my_any?
+    check = false
+    if block_given?
+      my_each { |i| check = true if yield(i) }
```
Comment (samgaco) [inline index.rb]: - [ ] my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: index.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    else
+      my_any? { |obj| obj }
+    end
+    check
+  end
+
+  def my_none?
+    check = true
+    if block_given?
+      my_each { |i| check = false if yield(i) }
```
Comment (samgaco) [inline index.rb]: - [ ] my_none? when no block or argument is given returns true only if none of the collection members is true
- [ ] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: index.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    else
+      my_none { |obj| obj }
+    end
+    check
+  end
+
+  def my_count(*arg)
+    count = 0
+    unless block_given?
+      if arg.empty?
```
Comment (samgaco) [inline index.rb]: - [ ] my_count returns the number of items in enum through enumeration


- [ ] my_count counts the number of items in enum that are equal to item if an argument is given

File: index.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      end
+    end
+    my_each { |i| count += 1 if yield(i) }
+    count
+  end
+
+  def my_map(&proc)
+    return arr.to_enum unless block_given?
+
+    result = []
```
Comment (samgaco) [inline index.rb:119]: - [ ] my_map returns an Enumerator if no block is given

File: index.rb [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    my_each { |i| result << proc.call(i) }
+    result
+  end
+
+  def my_inject(init = nil)
+    n = 0
+    arr = to_a
+    acc = init ? init + arr[0] : arr[0]
```
Comment (samgaco) [inline index.rb]: - [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block

```
irb(main):104:0> [1,2,3].my_inject(2) {|acc, n| acc * n}
=> 18
irb(main):105:0> [1,2,3].inject(2) {|acc, n| acc * n}
=> 12
irb(main):106:0> [1,2,3].my_inject(2) {|acc, n| acc + n}
=> 8
irb(main):107:0> [1,2,3].inject(2) {|acc, n| acc + n}
=> 8
irb(main):110:0> [1,2,3].inject(10) {|acc, n| acc * n}
=> 60
irb(main):111:0> [1,2,3].my_inject(10) {|acc, n| acc * n}
=> 66

```

- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

```
irb(main):099:0> [1,2,3].inject(:+)
=> 6
irb(main):100:0> [1,2,3].my_inject(:+)
Traceback (most recent call last):
....
NoMethodError (undefined method `+' for :+:Symbol)


```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @edemagbenyo :high_brightness: 

I left you some comments below about what functionalities your methods are lacking. Please contrast your methods against the original ones [here](https://ruby-doc.org/core-2.6.3/Enumerable.html).

If you would need further clarification with some concrete examples you may **ask me in the comments** below or slack, my name is _Samuel García_ over there.

This project is a bit tricky but it will help you with getting to know ruby better, so you will be feeling more fluent moving on, therefore good spirit :muscle: 

------------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: index.rb (hunk 5 above)
Comment (edemagbenyo) [reply index.rb]: Can you provide an example of how i should return the number of items through enumeration?

File: index.rb [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      else
+        my_each { |i| check = true if yield(i) == args[0] }
+      end
+      return check
+    elsif args.empty?
+      my_any? { |obj| obj }
+    else
+      my_any?(args[0]) { |obj| obj }
+    end
+  end
```
Comment (stratospherique) [inline index.rb]: * [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
* [ ] my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
* [ ] my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: index.rb [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      else
+        my_each { |i| check = false if yield(i) == args[0] }
+      end
+      return check
+    elsif args.empty?
+      my_none? { |obj| obj }
+    else
+      my_none?(args[0]) { |obj| obj }
+    end
+  end
```
Comment (stratospherique) [inline index.rb]: * [ ] my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
* [ ] my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
* [ ] my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: index.rb [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+      if arg.empty?
+        my_count { |i| i }
+      else
+        my_count { |i| arg[0] == i }
+      end
+    end
+    my_each { |i| count += 1 if yield(i) }
+    count
+  end
```
Comment (stratospherique) [inline index.rb]: * [ ] my_count returns the number of items in enum through enumeration
* [ ] my_count counts the number of items in enum that are equal to item if an argument is given

File: index.rb [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      my_each_with_index { |_, i| acc = yield(acc, arr[i]) }
+    elsif args[0].is_a?(Integer) && args[1].is_a?(Symbol)
+      acc = args[0]
+      my_each_with_index { |_, i| acc = args[1].to_proc.call(acc, arr[i]) }
+    elsif args[0].is_a?(Symbol)
+      my_each_with_index { |_, i| acc = args[0].to_proc.call(acc, arr[i]) }
+    end
+    acc
+  end
+end
```
Comment (stratospherique) [inline index.rb]: * [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block

File: index.rb (hunk 10 above)
Comment (stratospherique) [reply index.rb]: for Req.1 
```ruby
arr = new Array(100) {rand(0...9)}
puts arr.my_count === arr.count // returns false
```

for Req.2
```ruby
arr = [1,5,8,7,9,6,1,5]
p arr.count(1) === arr.my_count(1) // returns false
```

**Review by stratospherique — CHANGES_REQUESTED** [review]
### Great work implementing the suggested changes :clap: However, further changes needs to be made. check below :arrow_down: 

Check the code comments below for what requirements your methods should fulfill.

### that's it :relieved: When finished, kindly resubmit another code review request :pray: 

**Ahmed Mahfoudh**

File: index.rb [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+    result = []
+    my_each { |i| result << proc.call(i) }
+    result
+  end
+
+  def my_inject(*args)
+    arr = to_a
+    acc = 0
+    if block_given?
```
Comment (samgaco) [inline index.rb:127]: - [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block


this method keeps failing

```
irb(main):145:0> [1,2,3,4,5,6].my_inject {|n,s| n*s}
=> 0
irb(main):146:0> [1,2,3,4,5,6].inject {|n,s| n*s}
=> 720
```


```
irb(main):153:0> ["ff","wfrgrggwrg", "ggg","abc","exampleexampleexample", "nowis"].inject{ |memo, word| memo.length > word.length ? memo : word }
=> "exampleexampleexample"
irb(main):154:0> ["ff","wfrgrggwrg", "ggg","abc","exampleexampleexample", "nowis"].my_inject{ |memo, word| memo.length > word.length ? memo : word }
Traceback (most recent call last):
....
NoMethodError (undefined method `length' for 0:Integer)
```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @edemagbenyo :high_brightness: 

Your inject method keeps giving inaccurate results, please check the comments below and submit back for another code review once you fix it.

---------------------------

* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Review by whiz25 — APPROVED** [review]
## Awesome :clap: 
You have implemented all the requested changes and it's, therefore, a :+1:  from me. Go ahead and merge.

### Martin Afani

**Pull request merged**